### PR TITLE
Build rust crates that bind to C code

### PR DESCRIPTION
**Pull request by SchmErik**
This PR is the first step in enabling crates that are bindings to c code. This PR changes cargo-risczero to download a prebuilt c toolchain, and sets appropriate flags on risc0-build to invoke the c compiler when building guest crates. This solution is able to compile the blst crate so I've added a test case that exercises this crate. Note: while this does a signature verification, further testing is needed to conclude that BLST C code is working as expected by using utilities such as `cargo risczero test`.

### CODE REVIEW

**Comment by SchmErik** [comment]
@flaub this is ready for review. I used the 2024.01.05 toolchain and it allowed me to drop the script and the whole -I business. I hard-coded `cargo risczero install` to download this tag regardless of version for the time being.

File: risc0/build/src/lib.rs [hunk 1]
```diff
@@ -40,6 +40,21 @@ pub use docker::docker_build;
 
 const RUSTUP_TOOLCHAIN_NAME: &str = "risc0";
 
+/// todo
```
Comment (flaub) [inline risc0/build/src/lib.rs:43]: TODO

File: risc0/cargo-risczero/src/commands/install.rs [hunk 2]
```diff
@@ -82,10 +88,17 @@ impl Install {
...
+        &self,
+        toolchain_dir: &Path,
+    ) -> Result<(RustupToolchain, CToolchain)> {
         if let Some(target) = guess_host_target() {
-            match self.download_toolchain(target, toolchain_dir) {
-                Ok(path) => RustupToolchain::link(RUSTUP_TOOLCHAIN_NAME, &path.join("rust")),
+            match self.download_toolchains(target, toolchain_dir) {
+                Ok((rust_path, c_path)) => {
+                    let r = RustupToolchain::link(RUSTUP_TOOLCHAIN_NAME, &rust_path)?;
+                    let c = CToolchain::link(&c_path)?;
```
Comment (flaub) [inline risc0/cargo-risczero/src/commands/install.rs:99]: avoid single letter variable names

File: risc0/cargo-risczero/src/toolchain.rs [hunk 3]
```diff
@@ -15,16 +15,48 @@
...
+use fs_extra::dir::CopyOptions;
+use risc0_build::risc0_data;
 
 use crate::utils::CommandExt;
 
-/// Custom rust repository.
-pub const RUST_REPO: &str = "https://github.com/risc0/rust.git";
+pub enum ToolchainRepo {
+    Rust,
+    C,
```
Comment (flaub) [inline risc0/cargo-risczero/src/toolchain.rs:31]:
```suggestion
    Cpp,
```

File: risc0/cargo-risczero/src/toolchain.rs [hunk 4]
```diff
@@ -111,3 +143,50 @@ impl RustupToolchain {
         })
     }
 }
+
+/// A rustup toolchain manager
+#[derive(Clone, Debug)]
+pub struct CToolchain {
```
Comment (flaub) [inline risc0/cargo-risczero/src/toolchain.rs:149]:
```suggestion
pub struct CppToolchain {
```

File: risc0/cargo-risczero/src/toolchain.rs [hunk 5]
```diff
@@ -111,3 +143,50 @@ impl RustupToolchain {
...
+        let c_download_dir = Self::get_subdir(path)?;
+        let r0_data = risc0_data()?;
+        fs_extra::dir::copy(
+            c_download_dir.clone(),
+            &r0_data,
+            &CopyOptions::new().overwrite(true).copy_inside(true),
+        )?;
+
+        // for c, we will keep the toolchains in the r0_data directory for now
+        let c_install_dir = &r0_data.join("c");
```
Comment (flaub) [inline risc0/cargo-risczero/src/toolchain.rs:179]:
```suggestion
        let c_install_dir = &r0_data.join("cpp");
```

**Review by flaub — APPROVED** [review]
This is a good start. Some minor rename suggestions

**Comment by nazar-pc** [comment]
Does it resolve https://github.com/risc0/risc0/issues/328?

**Comment by Korolen** [comment]
Amazing!!

Is it compatible with risc0's deterministic builds mechanism ?

**Pull request merged**